torch/csrc/utils/tensor_new.cpp

```diff
@@ -546,11 +576,21 @@ Tensor new_tensor(const Type& type, PyObject* args, PyObject* kwargs) {
   ParsedArgs<4> parsed_args;
   auto r = parser.parse(args, kwargs, parsed_args);
   if (r.idx == 0) {
-    return new_from_data_copy(
+    PyObject* data = r.pyobject(0);
+    if (THPVariable_Check(data)) {
+      PyErr_WarnEx(PyExc_UserWarning,
+        "To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() "
```

Comment:
It seems okay to put the warning in internal_new_from_data (you can pass in the name of the function).  The problem with the previous implementation was that args_requires_grad wasn't respected across all code paths, but the name of the function can easily be respected across all code paths.

Reply:
Can I keep the warning at the caller function of `internal_new_from_data`? Otherwise I will need to put a default name at `internal_new_from_data` that might be accidentally picked up when a caller function triggers the warning..